subprojects/fflas-ffpack.wrap

```diff
@@ -0,0 +1,6 @@
+[wrap-file]
```

Comment:
You're scaring me, this looks a lot like a copy of sage-the-distro but within the new sagelib build system.

Reply:
I don't see why this all has to be a part of sagelib build. Preparing a chroot or something like this for sagelib to use for its dependencies has to be something separate.

Reply:
OTOH I'm OK with tolerating 10 or 20 such `subprojects/`.

Reply:
My concern is that once again it's impossible to work on sagelib without having to maintain a mini-distro of other people's software. That's 20 subprojects (with pinned versions) whose build systems can now fail as part of the sagelib build, and that need to be updated in sync with changes to sagelib instead of simply changing the version bound in `meson.build`. For subprojects that aren't available on any distro there might be some benefit to it, but all of sage's dependencies are available everywhere, and if (ever) not... we have sage-the-distro for that. Which hopefully somebody else will be maintaining while I work on sagelib for once.

Reply:
Yes, certainly, my concern too. However, there are cases where there is no correct dependency on your OS.
Are these dependencies conditional (it seems to be the case for GAP, but not for the other ones) ?

Reply:
The idea is to have them downloaded and bundled only if a system copy is not detected, exactly like sage-the-distro. But we know it's not that simple. If you use the bundled flint, for example, then you must also bundle anything that links to flint to avoid linking in multiple versions of the same symbols. The wrap system can't handle that because (same old story) it's not a real package manager and doesn't do dependency resolution and can't build anything not written in its own language.

Package management is simply better handled by conda/nix/guix or the package manager on your system or even sage-the-distro. But most importantly, it's better if _we_ don't have to handle bug reports and patches for 20 subprojects inside of sagelib all over again.

Reply:
> You're scaring me, this looks a lot like a copy of sage-the-distro but within the new sagelib build system.

Sorry, that was not my intention! I don't think that having a few subprojects that provide fallbacks in case the system-version is too old is a huge problem. My hope was that this would be limited to a few libraries (flint, gap, singular?). You have more experience with all the inter-dependencies. Do you think this list becomes too long, even if one only builds what is not provided by relatively recent linux distros? The goal was mostly to ease the migration to the meson setup for existing developers that might have not installed all dependencies, and to test it on more platforms (without any subprojects it only compiled on arch). It was definitely not meant as a mini sage-the-distro hell 😄. By the way, you can disable the fallback mechanism completely in meson.

But thanks both of you for your feedback. I'll remove all subprojects from this PR, and will continue my experiments at #37077.

Reply:
> > You're scaring me, this looks a lot like a copy of sage-the-distro but within the new sagelib build system.
> 
> Sorry, that was not my intention! I don't think that having a few subprojects that provide fallbacks in case the system-version is too old is a huge problem. My hope was that this would be limited to a few libraries (flint, gap, singular?, linbox? what else?). You have more experience with all the inter-dependencies. Do you think this list becomes too long, even if one only builds what is not provided by relatively recent linux distros? The goal was mostly to ease the migration to the meson setup for existing developers that might have not installed all dependencies, and to test it on more platforms (without any subprojects it only compiled on arch). It was definitely not meant as a mini sage-the-distro hell 😄. By the way, you can disable the fallback mechanism completely in meson.

And I'm sorry for my strong reaction, but I've been traumatized! I think it might be OK to have subprojects in _some_ cases.

Libraries generally cannot be bundled as subprojects because of the transitive linking issue. If meson accepts the system singular and if that singular links to the system flint-2.x, then building the flint-3.x subproject (because the system flint-2.x is not acceptable) won't suffice. When both are linked into sagelib there will be conflicts. This is why we have the `SAGE_SPKG_DEPCHECK` stuff in `spkg-configure.m4` and AFAIK meson doesn't have a way to handle it.  

Groups of packages with complicated version constraints are also out. Givaro, linbox, and fflas-ffpack are good examples of this. Their versions need to agree exactly, so bundling one while accepting the other two from the system can lead to incompatibilities.

Both of those are just instances of the one thing that I'd truly like to avoid though. Namely, having to deal with build failures in other peoples' packages while working on sagelib. If linbox fails to build as part of the sagelib build process, I don't want that to be my problem. And I don't want to be made responsible for sorting out the version constraints in the subprojects before we can loosen the linbox version constraint in meson.build. If all packages come from the same system, we can assume that they're consistent, and that saves us a lot of trouble.

So in light of all that there aren't a lot of packages that I think would make good subprojects. Actual subprojects of sage could be OK, since we're responsible for those bugs anyway. And any packages that are truly unavailable elsewhere, but I don't think we have any of those as dependencies right now, nor should we ever add them. Packages providing only executables and with uncomplicated dependencies would be the only other class that i can think of right now.

For what it's worth, flint is fixed in https://github.com/flintlib/flint/commit/e84002f3cc9c76a72a7ce1f8b834a02d89bd55c3 and I spent a week working on PALP to get it into Gentoo, so now this can be built "out of the box" on Gentoo as well, without the overlay. We're still missing a few data packages but the build doesn't fail because of it.

Reply:
> And I'm sorry for my strong reaction, but I've been traumatized! 

Thanks for opening up and sharing. And, no worries, we all are here for the same reason :smiley:

Jokes aside, thanks for the input and helping me realize that I went to deep, in fact, far deeper than I ever intended to.

Reply:
Removed these wrap files again.